force-app/main/default/classes/DMLOperations.cls

```diff
@@ -128,92 +107,130 @@ public with sharing class DMLOperations {
-     * @param oppNames The list of Opportunity names.
-     */
+    * Question 8
+    * Insert or updates opportunities from list of names.
+    * Associate the Opportunities with account that has the name accountName.
+    * There should be an Opportunity for each name provided in the list.
+    * @param accountName The name of the Account.
+    * @param oppNames The list of Opportunity names.
+    */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        Account acc = new Account();
+        List<Account> accList = [SELECT Id, Name FROM Account WHERE Name = :accountName LIMIT 1];
+        if(accList.isEmpty()) {
+            acc.Name = accountName;
+            insert acc;
+        } else {
+            acc = accList[0];
+        }
+
+        List<Opportunity> oppList = new List<Opportunity>();
+        List<Opportunity> existingOpps = [SELECT Id, Name FROM Opportunity WHERE Account.Name = :accountName];
+
+        for(String oppName : oppNames) {
+            Boolean opportunityExists = false;
+                for(Opportunity existingOpp : existingOpps) {
+                    if(existingOpp.Name == oppName) {
+                    opportunityExists = true;
+                    break;
+                    }
+                }
+
+            if(!opportunityExists) {
+                Opportunity newOpp = new Opportunity();
+                newOpp.AccountId = acc.Id;
+                newOpp.Name = oppName;
+                newOpp.StageName = 'New';
+                newOpp.CloseDate = Date.today().addMonths(3);
+                oppList.add(newOpp);
+            }
+        } 
+        upsert oppList;
     }
-    
+
     /**
      * Question 9
 	 * Upserts an Account with the specified name.
-	 * If an Account with the given name already exists, update the description to "Updated Account".
-	 * If no such Account exists, it creates a new one and sets the description to "New Account".
+	 * If an Account with the given name already exists, it updates that Account.
+	 * If no such Account exists, it creates a new one.
 	 * @param accountName The name of the Account to be upserted.
 	 * @return The Account that was upserted.
 	 */
 	public static Account upsertAccount(String accountName) {
-		// Query for the Account
-		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
-
-		// if the Account exists, update the description
-
-		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
+        Account acc;
+		List<Account> accList = [SELECT Id, Name FROM Account WHERE Name =: accountName];
+        if(!accList.isEmpty()) {
+            acc = accList[0];
+            acc.Description = 'Updated Account';
+        } else {
+            acc = new Account(Name = accountName);
+        }
```

Comment:
You're missing the Description field here (and this is why the test method failed):
```suggestion
        } else {
            acc = new Account(Name = accountName, Description = 'New Account');
        }
```